riscv-executor/src/lib.rs

```diff
@@ -776,65 +835,68 @@ mod builder {
             // fill up main trace to degree
             self.extend_rows(main_degree);
 
+            // generate witness for submachines
+            // ----------------------------
+            for (m, ops) in self.trace.submachine_ops {
+                let m = match m {
+                    MachineInstance::main_memory => "memory",
+                    MachineInstance::main_regs => "regs",
+                    MachineInstance::main_binary => "binary",
+                    MachineInstance::main_shift => "shift",
+                    MachineInstance::main_publics => "publics",
+                    MachineInstance::main_split_gl => "split_gl",
+                    MachineInstance::main_poseidon_gl => "poseidon_gl",
+                };
+                for op in ops {
```

Comment:
Could we somehow do this in parallel per machine?

Reply:
yes, was actually trying it out, doing it in a separate PR